document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/transformer/PostgresUnnestQueryTransformer.java

```diff
@@ -0,0 +1,267 @@
+import org.hypertrace.core.documentstore.expression.impl.KeyExpression;
+import org.hypertrace.core.documentstore.expression.impl.LogicalExpression;
+import org.hypertrace.core.documentstore.expression.impl.RelationalExpression;
+import org.hypertrace.core.documentstore.expression.impl.UnnestExpression;
+import org.hypertrace.core.documentstore.expression.operators.LogicalOperator;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.expression.type.FromTypeExpression;
+import org.hypertrace.core.documentstore.parser.FilterTypeExpressionVisitor;
+import org.hypertrace.core.documentstore.parser.FromTypeExpressionVisitor;
+import org.hypertrace.core.documentstore.parser.SelectTypeExpressionVisitor;
+import org.hypertrace.core.documentstore.query.Query;
+import org.hypertrace.core.documentstore.query.transform.QueryTransformer;
+import org.hypertrace.core.documentstore.query.transform.TransformedQueryBuilder;
+
+/**
+ * Accessing an item of a JSON array w/o index expression results in no match in Postgres.
+ *
+ * <p>As an example, { "item":"Soap", "sales":[ { "city":"delhi" } ] }. If we try to access
+ * `sales.city` it will return null as there is no key at `sales`.
+ *
+ * <p>However, as mongo supports -EQ- operator on array field that behaves as contains internally,
+ * `sales.city-EQ-delhi` is a valid filter expression. Ideally, the client should have expressed
+ * contains as `sales-CONTAINS-{"city":"delhi"}`.
+ *
+ * <p>To support compatibility, this pre-processor moves any applied filter on the array field to
+ * unnest expression filter.
+ */
+public class PostgresUnnestQueryTransformer implements QueryTransformer {
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public Query transform(Query query) {
+    // Don't do any thing if there is no unnest expression
+    if (query.getFromTypeExpressions().size() <= 0) {
+      return query;
+    }
+
+    // Prepare an AND tree of filters. If, any OR filter, the full sub-tree needs to move.
+    // [x, y, z, OR, OR]
+    AndFiltersCollector andFiltersCollector = new AndFiltersCollector();
+    List<FilterTypeExpression> andFilters =
+        query
+            .getFilter()
+            .map(
+                filterTypeExpression ->
+                    filterTypeExpression.<List<FilterTypeExpression>>accept(andFiltersCollector))
+            .orElse(List.of());
+
+    // check if any filter matches with unnest expression
+    // if matches, move the matched one to unnest filter and remove it from main filter list
+    List<FilterTypeExpression> finalAndFilters = new ArrayList<>();
+    List<FromTypeExpression> finalFromTypeExpressions =
+        matchUnnestExpressionsWithFilters(
+            query.getFromTypeExpressions(), andFilters, finalAndFilters);
+
+    // return same query if no change
+    if (andFilters.size() == finalAndFilters.size()) return query;
```

Comment:
Nit: Surround with curly braces even with single lines for improved readability.

Reply:
Done.